### PR TITLE
Dcmp improve

### PR DESCRIPTION
**Pull request by KnightKu**
Here are some minor improvements for dcmp.
"dcmp: improve the usage() 
dcmp: print the matched count of each conjunctions group in summary"
These two are used to improve the usage and summary log, in order to make it more friendly to users.

"dcmp: record the rawtime directly rather than converting from wtime"
This patch uses localtime() to record the time rather than MPI_Wtime (with additional conversion), to avoid some unexpected condition, e.g. time overflow.

"dcmp: skip the size differ check for dir"
This commit remove the size differ check for dir, as we are focusing the size check of files, not dirs, and on the other hand, size of dir may be different even contained files are the same, depends on the fs.

Please refer to the commits for details.

### CODE REVIEW

**Comment by sihara** [comment]
ping.. if someone review patches, that's much applicated!

File: src/common/mfu_flist.c [hunk 1]
```diff
@@ -684,6 +684,54 @@ uint64_t mfu_flist_file_get_perm(mfu_flist bflist, uint64_t idx) {
...
+    flist_t* flist = (flist_t*) bflist;
+    const char* filename = mfu_flist_file_get_name(flist, idx);
+    size_t val_bufsize = 1024;
+    void* val = (void*) MFU_MALLOC(val_bufsize);
+    ssize_t val_size;
+    int got_val = 0;
+
+    *acl_size = 0;
+
+    while(!got_val) {
```
Comment (dinatale2) [inline src/common/mfu_flist.c]: You can get rid of `got_val` if you make this a `while (retries < maxRetries)` loop and place a break statements in the `else` block that sets `got_val = 1`. You can then bound the loop with the retries to make sure the loop terminates.

File: src/common/mfu_flist.c [hunk 2]
```diff
@@ -684,6 +684,54 @@ uint64_t mfu_flist_file_get_perm(mfu_flist bflist, uint64_t idx) {
...
+                    type, filename, errno, strerror(errno)
+                   );
+                break;
+            }
+        } else {
+            if(val_size > 0 && val_bufsize == 0) {
+                val_bufsize = (size_t) val_size;
+                val = (void*) MFU_MALLOC(val_bufsize);
+            } else {
+                got_val = 1;
```
Comment (dinatale2) [inline src/common/mfu_flist.c]: Per the comment above about the `while` loop change, you can turn this block into:
```
*acl_size = val_size;
break;
```

File: src/dcmp/dcmp.c [hunk 3]
```diff
@@ -1768,7 +1861,8 @@ static void dcmp_disjunction_print(
 static int dcmp_disjunction_match(
     struct dcmp_disjunction* disjunction,
     strmap* map,
-    const char* key)
+    const char* key,
+    int is_src)
```
Comment (dinatale2) [inline src/dcmp/dcmp.c]: I believe that `stdbool.h` is included in this file. If `is_src` is being used as a boolean, consider using the `bool` type.

**Comment by KnightKu** [comment]
Hi @dinatale2 , thanks for you quick review, will refresh them as you commented soon, thanks.

**Comment by KnightKu** [comment]
Hi @dinatale2 ,could please review it again?
btw, could you also please help to review following patch on my tree?
https://github.com/KnightKu/fileutils/commit/49a8a94b96b7e11ca068d84e3974b7706cf3ab21
I am trying to add a progress option to dcp, so that it will print the progress timely, e.g. per 10s, but when I testing it with multiple threads, the tasks will block on MPI_Allreduce, then hang forever.

**Comment by dinatale2** [comment]
@KnightKu sorry I haven't reviewed this yet. I'll get you a review sometime early this week.

**Comment by sihara** [comment]
ping... if you can reveiw them, that's much helpful. thanks!

**Comment by KnightKu** [comment]
@dinatale2 ，@Adam could you please review them, these patches are landed on my local tree for a long time, and we tested them a lot, works well till now, we think they are useful to many other guys, so if you can review and merge them soon, that would be helpful, thanks.

**Comment by dsikich** [comment]
@KnightKu Hi, can you please send us all of the commands and/or tests you are running so that we can verify your output here?

**Comment by dsikich** [comment]
@KnightKu Thanks, so most of my output was the same. I did notice some edge cases:
```bash
./install/bin/dcmp /tmp/dir2 /tmp/dir1 -o SIZE=DIFFER:/tmp/xxxx
Fatal error in PMPI_Type_get_extent: Invalid datatype, error stack:
PMPI_Type_get_extent(121): MPI_Type_get_extent(MPI_DATATYPE_NULL, lb=0x7fffffffda10, extent=0x7fffffffda08) failed
PMPI_Type_get_extent(80).: Datatype for argument datatype is a null datatype
PMGR_COLLECTIVE ERROR: rank 0 on cab670: Called pmgr_abort() Code: 1, Msg: Fatal error in PMPI_Type_get_extent: Invalid datatype, error stack:
PMPI_Type_get_extent(121): MPI_Type_get_extent(MPI_DATATYPE_NULL, lb=0x7fffffffda10, extent=0x7fffffffda08) failed
PMPI_Type_get_extent(80).: Datatype for argument datatype is a null datatype
cab670{sikich1}49: ./install/bin/dwalk -i /tmp/xxxx -p
```
This happens when all of the files in both directories are the same size. So, we should add support for this case. 

I'm also having trouble parsing some of the options:

```bash
$: ./install/bin/dcmp /tmp/dir1/ /tmp/dir2/ -o EXIST=ONLY_SRC,TYPE=DIFFER,SIZE=DIFFER,UID=DIFFER,GID=DIFFER,PERM=DIFFER,MTIME=DIFFER
field "PERM" illegal
failed to parse expression "PERM=DIFFER"
failed to parse conjunction "PERM=DIFFER"

dcmp:161414 terminated with signal 11 at PC=40234d SP=7fffffffdb10.  Backtrace:
./install/bin/dcmp[0x40234d]
./install/bin/dcmp[0x406065]
./install/bin/dcmp[0x40630f]
./install/bin/dcmp[0x406cbe]
./install/bin/dcmp[0x406e49]
/lib64/libc.so.6(__libc_start_main+0xfd)[0x2aaaacf03d1d]
./install/bin/dcmp[0x402189]
```

I can successfully run the getfacl function:

```bash
cab670{sikich1}54: getfacl /tmp/dir2/file1.txt 
getfacl: Removing leading '/' from absolute path names
# file: tmp/dir2/file1.txt
# owner: sikich1
# group: sikich1
user::rw-
group::---
other::---
```

But I get a parsing error again here:
```bash 
cab670{sikich1}55: ./install/bin/dcmp /tmp/dir1/file1.txt /tmp/dir2/file1.txt -o ACL=DIFFER
field "ACL" illegal
failed to parse expression "ACL=DIFFER"
failed to parse conjunction "ACL=DIFFER"

dcmp:168303 terminated with signal 11 at PC=40234d SP=7fffffffdb40.  Backtrace:
./install/bin/dcmp[0x40234d]
./install/bin/dcmp[0x406065]
./install/bin/dcmp[0x40630f]
./install/bin/dcmp[0x406cbe]
./install/bin/dcmp[0x406e49]
/lib64/libc.so.6(__libc_start_main+0xfd)[0x2aaaacf03d1d]
./install/bin/dcmp[0x402189]
```

Also, did you also test these cases with multiple processes?

**Comment by KnightKu** [comment]
Hi @dsikich , yeah, sure, the above cmd&output is a quick show.
BTW, did you apply all the patches (perm and acl support ones)? I mean, the parsing fail only happened to unsupported options.

**Comment by dsikich** [comment]
@KnightKu I believe so, I pulled down the latest changes on master, then made a branch off of that with your pull request. The acl functions are in the branch I was testing on.

**Comment by KnightKu** [comment]
Hi @dsikich , sorry for late reply. But it is really strange,  I can not reproduce these issues.:(

**Comment by dsikich** [comment]
@KnightKu hmm ok, well I will check again. I'm primarily worried about this one:
```bash
./install/bin/dcmp /tmp/dir2 /tmp/dir1 -o SIZE=DIFFER:/tmp/xxxx
```
I will reproduce again, and give you the exact structure of the directory and/or files.

**Comment by dsikich** [comment]
@KnightKu can you provide me with your Lustre version, mpi you are using & version, and your version of SLURM?

Also, can you provide me with the contents of the directories you were using for this command:
```bash
./install/bin/dcmp /tmp/dir2 /tmp/dir1 -o SIZE=DIFFER:/tmp/xxxx
```
Since I am still getting this error when I run it:
```bash
Fatal error in PMPI_Type_get_extent: Invalid datatype, error stack:
PMPI_Type_get_extent(121): MPI_Type_get_extent(MPI_DATATYPE_NULL, lb=0x7fffffffd360, extent=0x7fffffffd358) failed
PMPI_Type_get_extent(80).: Datatype for argument datatype is a null datatype
PMGR_COLLECTIVE ERROR: rank 0 on cab1: Called pmgr_abort() Code: 1, Msg: Fatal error in PMPI_Type_get_extent: Invalid datatype, error stack:
PMPI_Type_get_extent(121): MPI_Type_get_extent(MPI_DATATYPE_NULL, lb=0x7fffffffd360, extent=0x7fffffffd358) failed
PMPI_Type_get_extent(80).: Datatype for argument datatype is a null datatype
```

**Comment by KnightKu** [comment]
Hi @dsikich , following is what I did, includes the version info.
`[root@server14 mpifileutils]# lctl lustre_build_version
Lustre version: 2.5.42.8-g2e588d8-PRISTINE-2.6.32-642.1.1.el6.x86_64
lctl   version: 2.5.42.8-g2e588d8-PRISTINE-2.6.32-642.1.1.el6.x86_64
[root@server14 mpifileutils]# mkdir /tmp/dir_src
[root@server14 mpifileutils]# mkdir /tmp/dir_dst
[root@server14 mpifileutils]# echo "abc" > /tmp/dir_src/file_foo
[root@server14 mpifileutils]# echo "abc" > /tmp/dir_dst/file_foo
[root@server14 mpifileutils]# mpirun --allow-run-as-root --version
mpirun (Open MPI) 1.8.1

Report bugs to http://www.open-mpi.org/community/help/
[root@server14 mpifileutils]# mpirun --allow-run-as-root ./install/bin/dcmp /tmp                                                                                        /dir_src  /tmp/dir_dst -o SIZE=DIFFER:/tmp/xxxx
Files which have different sizes: [0/0], dumped to "/tmp/xxxx"
`

btw, could you please provide the environment  you used to test? lustre version, mpi version and some others, so that I can try with them too, thanks.

**Comment by dsikich** [comment]
@KnightKu Okay, here is what I did:

```bash
cab668{sikich1}14: mkdir /tmp/dir_src
cab668{sikich1}15: mkdir /tmp/dir_dst
cab668{sikich1}16: echo "abc" > /tmp/dir_src/file_foo
cab668{sikich1}17: echo "abc" > /tmp/dir_dst/file_foo
cab1{sikich1}19: srun -N1 -n1 install/bin/dcmp /tmp/dir_src /tmp/dir_dst -o SIZE=DIFFER:/tmp/xxxx
Fatal error in PMPI_Type_get_extent: Invalid datatype, error stack:
PMPI_Type_get_extent(121): MPI_Type_get_extent(MPI_DATATYPE_NULL, lb=0x7fffffffd0a0, extent=0x7fffffffd098) failed
PMPI_Type_get_extent(80).: Datatype for argument datatype is a null datatype
srun: error: mvapich: rank 0: unexpected hostidlen = 260
srun: error: mvapich: Fatal error. Killing job
PMGR_COLLECTIVE ERROR: rank 0 on cab1: Called pmgr_abort() Code: 1, Msg: Fatal error in PMPI_Type_get_extent: Invalid datatype, error stack:
PMPI_Type_get_extent(121): MPI_Type_get_extent(MPI_DATATYPE_NULL, lb=0x7fffffffd0a0, extent=0x7fffffffd098) failed
PMPI_Type_get_extent(80).: Datatype for argument datatype is a null datatype
srun: forcing job termination
slurmd[cab1]: *** STEP 3730632.4 KILLED AT 2018-03-13T10:11:59 WITH SIGNAL 9 ***
srun: Job step aborted: Waiting up to 2 seconds for job step to finish.
srun: error: cab1: task 0: Exited with exit code 1
slurmd[cab1]: *** STEP 3730632.4 KILLED AT 2018-03-13T10:11:59 WITH SIGNAL 9 ***
```

We cannot run jobs as root here. What does --allow-run-as-root exactly do?

Lustre version:
lustre-2.5.5-13chaos_2.6.32_696.20.1.1chaos.ch5.6.x86_64.x86_64

And I am using mvapich2, but I also tried running with mpirun and openmpi (1.8) with the same result:

``` bash
cab1{sikich1}25: mpirun -n 1 install/bin/dcmp /tmp/dir_src /tmp/dir_dst -o SIZE=DIFFER:/tmp/xxxx
Fatal error in PMPI_Type_get_extent: Invalid datatype, error stack:
PMPI_Type_get_extent(121): MPI_Type_get_extent(MPI_DATATYPE_NULL, lb=0x7fffffffcbf0, extent=0x7fffffffcbe8) failed
PMPI_Type_get_extent(80).: Datatype for argument datatype is a null datatype
PMGR_COLLECTIVE ERROR: rank 0 on cab1: Called pmgr_abort() Code: 1, Msg: Fatal error in PMPI_Type_get_extent: Invalid datatype, error stack:
PMPI_Type_get_extent(121): MPI_Type_get_extent(MPI_DATATYPE_NULL, lb=0x7fffffffcbf0, extent=0x7fffffffcbe8) failed
PMPI_Type_get_extent(80).: Datatype for argument datatype is a null datatype
-------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code.. Per user-direction, the job has been aborted.
-------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[43218,1],0]
  Exit code:    1
--------------------------------------------------------------------------


```

**Comment by sihara** [comment]
@dsikich we couldn't sucess your problem yet, but i want to reproduce and check your problem.
cab668{sikich1}14: mkdir /tmp/dir_src
cab668{sikich1}15: mkdir /tmp/dir_dst
cab668{sikich1}16: echo "abc" > /tmp/dir_src/file_foo
cab668{sikich1}17: echo "abc" > /tmp/dir_dst/file_foo
cab1{sikich1}19: srun -N1 -n1 install/bin/dcmp /tmp/dir_src /tmp/dir_dst -o SIZE=DIFFER:/tmp/xxxx

what is cab668 and cab1? is that hostname? are you sure running dcmp on node has /tmp/dir_src and /tmp/dir_dst? e.g. you created /tmp/dir_src and /tmp/dir_dst on cab668, but run dcmp on cab1?

**Comment by dsikich** [comment]
@sihara I realized the issue yesterday. I was rebuilding the code with ./buildme instead of ./buildme_dev so it was using an old version of libmfu out of the install directory instead of the new src/.

**Comment by dsikich** [comment]
@sihara @KnightKu @dinatale2 In any case, before merging I'd like a better description of what the code does in both the pull request and the commit messages. For instance, the "add ACL comparison support" commit. It would help if the pull request also listed all of the major changes and why they were added. I think splitting up the ACL comparison work with the summary improvements in dcmp would also speed up the merge process if you want to do that.

**Comment by sihara** [comment]
@dsikich thanks for your checking and hope you could move forward your testing of patch. And I agreed with you adding more descriptions to what/why we changed with patch. we will improve this sooner.

**Comment by dsikich** [comment]
@sihara Great, yes I was able to move forward. Thanks for all of your work.

**Comment by KnightKu** [comment]
@dsikich ，yeah, indeed, more description for each patch is needed, will improve it as you suggested, and spit this series into detail topics, to make them easy for you guys review and move forward. Anyway, thanks very much.

**Comment by KnightKu** [comment]
Hi @dsikich , I have updated the changelog of patches and comment of the PR, and split the ACL comparison patch into another PR (not pushed yet), so hope you guys can review this one first and move forward, thanks.

**Comment by KnightKu** [comment]
Hi guys, could you please have a look and move forward? any comment is welcome, thanks.

**Comment by dsikich** [comment]
@KnightKu What do you mean by this in your PR description?
"This patch uses localtime() to record the time rather than MPI_Wtime (with additional conversion), to avoid some unexpected condition."

What was the unexpected condition? Also, the first three commits look good and are fairly straightforward. 

I'd like to have at least two people look at the last one (dcmp: skip the size differ check for dir). If you pull that one into it's own PR. We can merge the summary and timing commits.

**Comment by KnightKu** [comment]
@dsikich anyway, thanks, let me refresh it.

**Comment by sihara** [comment]
hello, Just wanted to ping. would you re-review patch and commit comments please?

File: src/dcmp/dcmp.c [hunk 4]
```diff
@@ -1871,11 +1892,30 @@ static int dcmp_output_flist_match(
...
+        }
+        else {
+            uint64_t dst_matched_tmp;          /* matched dst items in this conjunction */
+            dst_matched_tmp = conjunction->dst_matched;          /* matched dst items in this conjunction */
+
+            MPI_Allreduce(&dst_matched_tmp, &(conjunction->dst_matched), 1,
+                          MPI_UINT64_T, MPI_SUM, MPI_COMM_WORLD);
+        }
+    }
     return 0;
```
Comment (dsikich) [inline src/dcmp/dcmp.c]: @KnightKu If you use an flist for src and dst matched, then we can remove this MPI code in the tool. That is how we also keep track of our src_compare and dst_compare lists for syncing directories (in dcmp_strmap_compare in dcmp.c for an example). If you add the matched items to the flist, then you can do a "mfu_flist_global_size." I think this would be simpler.

Comment (dsikich) [reply src/dcmp/dcmp.c]: @KnightKu We try to keep MPI calls in the common library, and out of the tools themselves.

Comment (KnightKu) [reply src/dcmp/dcmp.c]: Hi @dsikich , yeah, agreed, let's follow code conventions, will update it soon, thanks.

**Comment by KnightKu** [comment]
Hi @dsikich , could please review it again? thanks
